### ISSUE TITLE
Multiple values for a same parameter in query string no longer works

### ISSUE THREAD

**Issue by ippeiukai** [sneas/ionic-native-http-connection-backend#99]
With #85 (https://github.com/sneas/ionic-native-http-connection-backend/pull/85/files#diff-2ae3f38181854c596f6444c3e6224b09L79), it seems `ionic-native-http-connection-backend` since 6.2 is no longer sending all the param values in HttpParams. Specifically, when multiple values for a same parameter is specified, only one value is specified to that parameter in the request.

Perhaps related to:
https://github.com/silkimen/cordova-plugin-advanced-http/issues/242

**Comment by ippeiukai** [comment]
Actually, this probably is the cause:
https://github.com/sneas/ionic-native-http-connection-backend/pull/85/files#diff-2ae3f38181854c596f6444c3e6224b09R201

Nevertheless, cordova-plugin-advanced-http does not seem to support it.
Type def of `@ionic-native/http` does not seem to accept multiple values either:
https://github.com/ionic-team/ionic-native/blob/v5.19.1/src/@ionic-native/plugins/http/index.ts#L342

**Comment by sneas** [comment]
Hi, thank you for opening the issue. Could you please show the `HttpClient` call replicating the problem?

**Comment by ippeiukai** [comment]
I've copied and modified the relevant test cases to replicate the problem:

```TS
    it(`specifies correct params for get requests when a key has multiple values`, done => {
        const request = new HttpRequest('GET', 'http://test.com', {
            params: new HttpParams().append('a[]', '1').append('a[]', '2'),
        });
        expect(request.urlWithParams).toEqual('http://test.com?a%5B%5D=1&a%5B%5D=2');

        const spy = spyOn(http, 'sendRequest').and.returnValue(
            Promise.resolve({
                status: 200,
                data: {},
            }),
        );
        httpBackend.handle(request).subscribe(() => {
            expect(http.sendRequest).toHaveBeenCalledWith('http://test.com', {
                params: { 'what is expected?': 1 },
                method: 'get',
                serializer: 'urlencoded',
                headers: {},
                responseType: 'text',
            });
            done();
        });
    });
```

```TS

    it('posts http-param body with a key having multiple values', done => {
        const httpParamBody = new HttpParams()
            .append('a[]', '1')
            .append('a[]', '2');
        expect(httpParamBody.toString()).toEqual('a%5B%5D=1&a%5B%5D=2');

        const request = new HttpRequest(
            'POST',
            'http://test.com',
            httpParamBody,
            {
                headers: new HttpHeaders({
                    'Content-Type': ['application/x-www-form-urlencoded'],
                }),
            },
        );

        spyOn(http, 'sendRequest').and.returnValue(
            Promise.resolve({
                status: 200,
                data: '{}',
                headers: {},
            }),
        );

        httpBackend.handle(request).subscribe(() => {
            expect(http.sendRequest).toHaveBeenCalledWith(
                expect.anything(),
                expect.objectContaining({
                    data: { 'what is expected?': 1 },
                    serializer: 'urlencoded',
                }),
            );
            done();
        });
    });
```

**Comment by ippeiukai** [comment]
This may or may not be relevant, but there seems to be another bug here:

```TS

    it('posts http-param body with urlencoded not json automatically', done => {
        const httpParamBody = new HttpParams().append('a', '1');
        const request = new HttpRequest(
            'POST',
            'http://test.com',
            httpParamBody,
        );
        expect(request.detectContentTypeHeader()).toMatch(
            /^application\/x-www-form-urlencoded/,
        );

        spyOn(http, 'sendRequest').and.returnValue(
            Promise.resolve({
                status: 200,
                data: '{}',
                headers: {},
            }),
        );

        httpBackend.handle(request).subscribe(() => {
            expect(http.sendRequest).toHaveBeenCalledWith(
                expect.anything(),
                expect.objectContaining({
                    data: { a: 1 },
                    serializer: 'urlencoded', // fails with 'json'
                }),
            );
            done();
        });
    });
```

**Comment by ippeiukai** [comment]
🤔 Looking at the code, I suspect HttpHeaders have a similar problem with multiple value:.

collectionToObject extracts only single value per key:
https://github.com/sneas/ionic-native-http-connection-backend/blob/v6.2.1/src/utils/collection-to-object.ts#L14

HttpHeaders can have multiple values per key and should be represented as a comma-separated string in the raw request:
https://angular.io/api/common/http/HttpHeaders#getAll
https://github.com/angular/angular/blob/8.2.14/packages/common/http/src/xhr.ts#L93
https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2

**Comment by sneas** [comment]
Thank you for the helpful example, @ippeiukai ! I'll have a look into it whenever possible.

**Comment by sneas** [comment]
Hi @ippeiukai, thanks again for the issue and useful test cases.

Another issue regarding incorrectly detected serializer has been created.

**Comment by ippeiukai** [comment]
@sneas Thank you for taking a look!

This fix is useful in some situation, but I think the behaviour is different from the angular’s default backend. I haven’t actually tried the new version, but having looked at cordova-plugin-advanced-http code yesterday, I believe the request ends up as follows:

- `new HttpParams().append('a', '1').append('a', '2').toString()`: `a=1&a=2`
- request sent: `?a%5B%5D=1&a%5B%5D=2`

Basically, cordova-plugin-advanced-http seems to encode arrays and objects in query string in a particular way. I wonder if we can specify a string or an ArrayBuffer to params/data, so that serializing HttpParams can be handled in JS before passing it to native world.

**Comment by sneas** [comment]
Hi @ippeiukai, I see what you're talking about. I need to do some investigations to take care of this.

**Comment by sneas** [comment]
Hi, @ippeiukai. The fix has been applied for version 7.x.x which works with Angular >9 and Ivy renderer. Please let me know if you need the fix in version 6.x.x and I'll provide it in the separate release channel.

**Comment by ippeiukai** [comment]
We are yet to try Ionic 5 let alone Ivy. Backport release would be nice.
Thanks for your work!
